Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean

```diff
@@ -0,0 +1,648 @@
+    (I : Ideal A) :
+  (I.map (Algebra.TensorProduct.includeLeft : A →ₐ[R] A ⊗[R] B)).carrier =
+    LinearMap.range (LinearMap.rTensor B (Submodule.subtype (I.restrictScalars R))) := by
+  rw [← Ideal.map_includeLeft_eq]
+  rfl
+
+/-- The ideal of A ⊗[R] B generated by I is the image of A ⊗[R] I -/
+def Ideal.map_includeRight (I : Ideal B) : Ideal (A ⊗[R] B) := {
+  carrier := LinearMap.range (LinearMap.lTensor A (Submodule.subtype (I.restrictScalars R)))
+  add_mem' := fun {x} {y} hx hy  ↦ Submodule.add_mem _ hx hy
+  zero_mem' := by
+    simp only [SetLike.mem_coe]
+    apply zero_mem
+  smul_mem' := fun x y ↦ by
+    simp only [SetLike.mem_coe]
+    rintro ⟨y, rfl⟩
+    induction x using TensorProduct.induction_on with
+    | zero =>
+        rw [zero_smul]
+        apply zero_mem
+    | tmul a b =>
+        induction y using TensorProduct.induction_on with
+        | zero =>
+            rw [map_zero]
+            apply zero_mem
+        | tmul x y =>
+            simp only [LinearMap.lTensor_tmul, Submodule.coeSubtype, smul_eq_mul,
+              Algebra.TensorProduct.tmul_mul_tmul]
+            suffices h : b * ↑y ∈ Submodule.restrictScalars R I
+            use (a * x) ⊗ₜ[R] ⟨b * ↑y, h⟩
+            rfl
+            simp only [Submodule.coe_restrictScalars, Submodule.restrictScalars_mem]
+            apply Ideal.mul_mem_left
+            exact y.prop
+        | add x y hx hy =>
+            rw [map_add, smul_add]
+            exact Submodule.add_mem _ hx hy
+    | add x x' hx hx' =>
+        rw [add_smul]
+        exact Submodule.add_mem _ hx hx' }
+
+/-- The ideal of A ⊗[R] B generated by I is the image of A ⊗[R] I -/
+lemma Ideal.map_includeRight_eq (I : Ideal B) :
+  (I.map_includeRight : Ideal (A ⊗[R] B)) =
+      I.map (Algebra.TensorProduct.includeRight : B →ₐ[R] A ⊗[R] B) := by
+  apply le_antisymm
+  · rintro x ⟨y, rfl⟩
+    induction y using TensorProduct.induction_on with
+    | zero =>
+        rw [map_zero]
+        apply zero_mem
+    | tmul a b =>
+        simp only [LinearMap.lTensor_tmul, Submodule.coeSubtype]
+        suffices : a ⊗ₜ[R] (b : B) = (a ⊗ₜ[R] (1 : B)) * ((1 : A) ⊗ₜ[R] (b : B))
+        rw [this]
+        apply Ideal.mul_mem_left
+        apply Ideal.mem_map_of_mem
+        exact Submodule.coe_mem b
+        simp only [Submodule.coe_restrictScalars, Algebra.TensorProduct.tmul_mul_tmul,
+          mul_one, one_mul]
+    | add x y hx hy =>
+        rw [map_add]
+        apply Ideal.add_mem _ hx hy
+  · rw [Ideal.map, Ideal.span_le]
+    rintro x ⟨b, hb, rfl⟩
+    simp only [Algebra.TensorProduct.includeRight_apply, Submodule.coe_set_mk,
+      AddSubmonoid.coe_set_mk, AddSubsemigroup.coe_set_mk, SetLike.mem_coe, LinearMap.mem_range]
+    use 1 ⊗ₜ[R] ⟨b, hb⟩
+    rfl
+
+/-- The ideal of A ⊗[R] B generated by I is the image of A ⊗[R] I -/
+lemma Algebra.TensorProduct.ideal_map_includeRight_carrier
+    {R : Type _} [CommSemiring R]
+    {A B : Type _} [Semiring A] [Semiring B] [Algebra R A] [Algebra R B]
```

Comment:
I can never remember whether we're supposed to be using `Type*` or `Type _` nowadays, but I suspect we're not supposed to be using both. I think it's `Type*` right?

Reply:
`Type*` but when I wrote this, it was in the middle of `Type*` being promoted.